os/lib/iq-seeder.c

```diff
@@ -0,0 +1,161 @@
+ */
+
+#include "lib/iq-seeder.h"
+#include "net/netstack.h"
+#include "lib/aes-128.h"
+#include <string.h>
+
+#ifdef IQ_SEEDER_CONF_SAMPLING_RATE
+#define SAMPLING_RATE IQ_SEEDER_CONF_SAMPLING_RATE
+#else /* IQ_SEEDER_CONF_SAMPLING_RATE */
+#define SAMPLING_RATE 200 /* Hz */
+#endif /* IQ_SEEDER_CONF_SAMPLING_RATE */
+
+#define COLUMN_COUNT  50
+#define ROW_COUNT     16
+
+static const uint8_t toeplitz[COLUMN_COUNT + ROW_COUNT - 1] =
+    {  93 ,  50 , 210 , 134 ,  79 ,  52 , 237 , 192 ,  40 , 201 ,
+        3 , 184 , 152 ,  74 ,  27 ,  28 ,  32 , 111 ,  79 , 222 ,
+      174 ,  51 , 223 ,  66 , 152 , 211 , 234 , 124 ,  92 ,  64 ,
+      206 , 169 , 227 , 155 , 106 ,  87 , 207 , 135 , 238 , 101 ,
+      254 , 163 ,  55 ,  76 ,  50 ,  40 ,   4 , 149 ,  27 ,   1 ,
+      127 , 159 , 160 ,  91 , 251 , 179 , 186 , 200 , 225 ,  47 ,
+      235 , 223 ,  39 , 117 ,  19 };
+
+/*---------------------------------------------------------------------------*/
+static uint8_t
+get_toeplitz_element(uint8_t row, uint8_t column)
+{
+  uint8_t min;
+
+  min = row < column ? row : column;
+  row -= min;
+  column -= min;
+
+  return toeplitz[row ? COLUMN_COUNT - 1 + row : column];
+}
+/*---------------------------------------------------------------------------*/
+/* from http://en.wikipedia.org/wiki/Finite_field_arithmetic */
+static uint8_t
+mul_gf_256(uint8_t a, uint8_t b)
+{
+  uint8_t p;
+  uint8_t i;
+
+  p = 0;
+  for(i = 0; i < 8; i++) {
+    if(b & 1) {
+      p ^= a;
+      a <<= 1;
+      if(a & 0x100) {
+        a ^= 0x11b;
+      }
+      b >>= 1;
+    }
+  }
+  return p;
+}
+/*---------------------------------------------------------------------------*/
+static void
+extract(uint8_t *target, uint8_t *source)
+{
+  uint8_t row;
+  uint8_t column;
+
+  /**
+   * Toeplitz matrix-based extractor. For theory, see "True Random Number
+   * Generators Secure in a Changing Environment: Improved Security Bounds"
+   * by Skorski.
+   */
+  for(row = 0; row < ROW_COUNT; row++) {
+    target[row] = 0;
+    for(column = 0; column < COLUMN_COUNT; column++) {
+      target[row] ^= mul_gf_256(get_toeplitz_element(row, column), source[column]);
+    }
+  }
+}
+/*---------------------------------------------------------------------------*/
+static void
+seed_16_bytes(uint8_t *result)
+{
+  uint8_t bit_pos;
+  uint8_t byte_pos;
+  uint16_t iq_count;
+  uint8_t accumulator[COLUMN_COUNT];
+  radio_value_t iq;
+
+  bit_pos = 0;
+  byte_pos = 0;
+  memset(accumulator, 0, COLUMN_COUNT);
+
+  NETSTACK_RADIO.on();
+  for(iq_count = 0; iq_count <= (COLUMN_COUNT * 8 / 2); iq_count++) {
+    NETSTACK_RADIO.get_value(RADIO_PARAM_IQ_LSBS, &iq);
+
+    if(!iq_count) {
+      /* drop first I/Q reading because it is always 0 for some reason */
+      continue;
+    }
+
+    /* append I/Q LSBs to accumulator */
+    accumulator[byte_pos] |= iq << bit_pos;
+    bit_pos += 2;
+    if(bit_pos == 8) {
+      bit_pos = 0;
+      byte_pos++;
+    }
+  }
+  NETSTACK_RADIO.off();
```

Comment:
If I got you right, you are concerned that this may interfere with the MAC protocol. However, currently, seeding just happens once at start up before the network stack is initialized. For implementing reseeding at runtime, my plan would be to add a function to `csprng.c`, which gets called if a new seed is ready to be mixed with the current seed. In this way, the MAC protocol can, e.g., collect I/Q data in the background and notify the CSPRNG without any conflicts.

Reply:
I'm with you now, yep. So this happens only once as you say, and that's before the stack gets initialised. All good.

Reply:
Maybe a comment in the code explaining that would be a good, for users to understand the limitations.

Reply:
I am now mentioning this limitation in the doxygen description of `iq_seeder_seed`